src/js/client/editprefset-viewport.js

```diff
@@ -0,0 +1,1001 @@
+        openAddContextDialog: null,
+        openEditContextDialog: null,
+        openConfirmDialog: null,
+        openConfirmAddProductDialog: null,
+        openConfirmDeleteContextDialog: null,
+        openConfirmSaveDialog: null,
+        openConfirmRemoveProductDialog: null,
+        startDevWidgets: null
+    },
+    bindings: {
+    },
+    components: {
+        // TESTING
+        baseDialog: {
+            type: "gpii.devpmt.dialogs.baseDialog",
+            createOnEvent: "openBaseDialog",
+            container: "{that}.dom.modalDialogContainer",
+            options: {
+                selectors: {
+                    initial: "#pmt-modal-dialog-render"
+                }
+            }
+        },
+        // TESTING
+        confirmationDialog: {
+            type: "gpii.devpmt.dialogs.confirmDialog",
+            createOnEvent: "openConfirmDialog",
+            container: "{that}.dom.modalDialogContainer",
+            options: {
+                selectors: {
+                    initial: "#pmt-modal-dialog-render"
+                }
+            }
+        },
+        confirmAddProductDialog: {
+            type: "gpii.devpmt.dialogs.confirmAddProductDialog",
+            createOnEvent: "openConfirmAddProductDialog",
+            container: "{that}.dom.modalDialogContainer",
+            options: {
+                selectors: {
+                    initial: "#pmt-modal-dialog-render"
+                },
+                model: {
+                    appId: "{gpii.devpmt.editPrefs}.model.activeModalDialog.appId",
+                    name: "{gpii.devpmt.editPrefs}.model.activeModalDialog.name"
+                }
+            }
+        },
+        confirmRemoveProductDialog: {
+            type: "gpii.devpmt.dialogs.confirmRemoveProductDialog",
+            createOnEvent: "openConfirmRemoveProductDialog",
+            container: "{that}.dom.modalDialogContainer",
+            options: {
+                selectors: {
+                    initial: "#pmt-modal-dialog-render"
+                },
+                model: {
+                    appId: "{gpii.devpmt.editPrefs}.model.activeModalDialog.appId",
+                    name: "{gpii.devpmt.editPrefs}.model.activeModalDialog.name",
+                    product: "{gpii.devpmt.editPrefs}.model.activeModalDialog.product",
+                    context: "{gpii.devpmt.editPrefs}.model.activeModalDialog.context"
+                }
+            }
+        },
+        addContextDialog: {
+            type: "gpii.devpmt.dialogs.addContextDialog",
+            createOnEvent: "openAddContextDialog",
+            container: "{that}.dom.modalDialogContainer",
+            options: {
+                selectors: {
+                    initial: "#pmt-modal-dialog-render"
+                },
+                model: {
+                    contextNames: "{gpii.devpmt.editPrefs}.model.contextNames",
+                    flatPrefs: "{gpii.devpmt.editPrefs}.model.flatPrefs"
```

Comment:
Looking at `gpii.devpmt.dialogs.addContextDialog`, adding contexts doesn't seem requiring `flatPrefs` in the model.

Reply:
Sorry, you do need `flatPrefs` at adding contexts. But you are still reading this value from the parent component [`gpii.devpmt.editPrefs`](https://github.com/GPII/gpii-devpmt/pull/1/files#diff-39a204f7a5b7809d23e44e2e8e91d27dR73) rather than this relayed value. One goal of the model relay is to make subcomponents more self-contained. Reaching out to outer components defeats this purpose. Please fix `gpii.devpmt.dialogs.addContextDialog` to read this value from the component itself rather than the parent. Thanks.

The same comment applies to other places such as `gpii.devpmt.dialogs.editContextDialog`, `gpii.devpmt.dialogs.confirmDeleteContextDialog` etc.

Reply:
I've more or less moved all of these to model relays, although I've run into a few issues that only seem to manifest when using the model relayed change applier. I'm trying to track down the central issue of this.